### PR TITLE
Add support for setting HDF5 alignment property when creating a file

### LINKED ISSUE #2177: Allow users to specify data alignment
In the mailing list I raised the question about data being aligned.

I provided a small code example that showed how data could be misaligned (code in python with a recent version of netcdf4-python)
<details> <summary> Code Example </summary>

```python
import xarray as xr
import numpy as np
import netCDF4
from pathlib import Path

basic_filename = "basic_file_netcdf4.nc"
if Path(basic_filename).exists():
    Path(basic_filename).unlink()

dataset = xr.DataArray(
    np.zeros((3072, 3072), dtype='uint8'),
    dims=("y", "x"),
    coords={
        "y": np.arange(3072, dtype=int),
        "x": np.arange(3072, dtype=int),
    },
    name='images').to_dataset()

dataset.to_netcdf(basic_filename, format="NETCDF4", engine="netcdf4")

import h5py
h5file = h5py.File(basic_filename)
h5dataset = h5file.get("images")
offset = h5dataset.id.get_offset()
print(offset % 4096)
print(offset % 2048)
print(offset % 1024)
print(offset % 512)
print(offset % 128)
print(offset % 64)

"""
3206
1158
134
134
6
6
"""
```
</details>

We came to the point where we determined that he user should set the File Access Property at Creation time, near:
https://github.com/Unidata/netcdf-c/blob/988e771a9ed99619c2e3261aea81f127dd7fa3d8/libhdf5/hdf5open.c#L772

I'm opening this as an issue to keep track of the conversation on github.

cc: @edhartnett 

Mailing list link: https://www.unidata.ucar.edu/mailing_lists/archives/netcdfgroup/2022/msg00000.html

### PR DESCRIPTION
**Pull request by DennisHeimbigner**
* Fixes #2177 
* Fixes #2178

re: https://github.com/Unidata/netcdf-c/issues/2177
re: https://github.com/Unidata/netcdf-c/pull/2178

Provide get/set functions to store global data alignment
information and apply it when a file is created.

The api is as follows:
````
int nc_set_alignment(int threshold, int alignment);
int nc_get_alignment(int* thresholdp, int* alignmentp);
````

If defined, then for every file created opened after the call to
nc_set_alignment, for every new variable added to the file, the
most recently set threshold and alignment values will be applied
to that variable.

The nc_get_alignment function return the last values set by
nc_set_alignment.  If nc_set_alignment has not been called, then
it returns the value 0 for both threshold and alignment.

The alignment parameters are stored in the NCglobalstate object
(see below) for use as needed. Repeated calls to nc_set_alignment
will overwrite any existing values in NCglobalstate.

The alignment parameters are applied in libhdf5/hdf5create.c
and libhdf5/hdf5open.c

The set/get alignment functions are defined in libsrc4/nc4internal.c.

A test program was added as nc_test4/tst_alignment.c.

## Misc. Changes Unrelated to Alignment

* The NCRCglobalstate type was renamed to NCglobalstate to
  indicate that it represented more general global state than
  just .rc data.  It was also moved to nc4internal.h.  This led
  to a large number of small changes: mostly renaming. The
  global state management functions were moved to nc4internal.c.

* The global chunk cache variables have been moved into
  NCglobalstate.  As warranted, other global state will be moved
  as well.

* Some misc. problems with the nczarr performance tests were corrected.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
Is it guaranteed that the functions are "atomic" and threadsafe?

**Comment by DennisHeimbigner** [comment]
No, but no part of the netcdf-c library has those properties.

**Comment by hmaarrfk** [comment]
Understood. Thank you for the clarification.

**Comment by Dave-Allured** [comment]
```
nc_set_alignment
nc_get_alignment
```
Are these to be public facing functions in the public netcdf-c API?

If yes, then I suggest add a standard documentation block for each function.  The above description block on this PR is a good start.

**Comment by DennisHeimbigner** [comment]
They are not part of the dispatch table since they are global. In this, they are like
the cache functions.

**Comment by Dave-Allured** [comment]
> They are not part of the dispatch table since they are global. In this, they are like
> the cache functions.

That seems reasonable to me.  For file types other than netcdf-4, these nc_*_alignment functions will behave as no-ops.

**Comment by Dave-Allured** [comment]
@DennisHeimbigner, thank you for adding the function docs.

Initially I found HDF5 disk alignment to be confusing in several ways.  I took the liberty of drafting some detail documentation to address the nuances.  Please consider these extra suggestions for `nc_set_alignment`.  Feel free to accept any parts and reword as you see fit.
.........................
`nc_set_alignment` causes new data written to a netCDF-4 file to be aligned on disk to a specified block size.  To be effective, `alignment` should be the system disk block size, or a multiple of it.  This setting is effective with MPI I/O and other parallel systems, not with single process file access.

This is a trade-off of write speed versus file size.  Alignment leaves holes between file objects.  The default of no alignment writes file objects contiguously, without holes.  Alignment has no impact on file readability.

Alignment settings apply only indirectly, through the file open functions.  Call `nc_set_alignment first`, then `nc_create` or `nc_open` for one or more files.  Current alignment settings are locked in when each file is opened, then forgotten when the same file is closed.   For illustration, it is possible to write different files at the same time with different alignments, by interleaving `nc_set_alignment` and `nc_open` calls.

Alignment applies to all newly written low-level file objects at or above the `threshold` size, including chunks of variables, attributes, and internal infrastructure.  Alignment is not locked in to a data variable.  It can change between data chunks of the same variable, based on a file's history.

Refer to [H5Pset_alignment](https://portal.hdfgroup.org/display/HDF5/H5P_GET_ALIGNMENT) in HDF5 documentation for more specific details, interactions, and additional rules.

**Comment by Dave-Allured** [comment]
I should probably have done this as a pull request to your branch.  Let me know if you prefer that.

**Comment by hmaarrfk** [comment]
I would like to suggest removing the phrase

> not with single process file access.

In fact, my initial request was centered around optimizing single process access time to files of mine.

I'm not adapt enough to provide hard examples with C++, but with the standard strategies, I reach 1GB/s read/write throughput.

With data alignment, and the HDF5 Direct driver, I can reach 3GB/s (NVMe limited speeds) on read and write. Thus, I would disagree that data alignment does not help with "single process" access.

Generally, it allows lower level hardware/firmware/software to organize data transfers in a highly effective way.

**Comment by hmaarrfk** [comment]
I had an other writeup of some experiments of mine: https://github.com/pydata/xarray/issues/6309

Eventually, xarray uses netcdf4-python as a backend.

**Comment by Dave-Allured** [comment]
@hmaarrfk, it is good to work out these issues up front.  Thanks for commenting.

Those notes about single and parallel performance were taken straight from the HDF5 function documentation for [H5Pset_alignment](https://portal.hdfgroup.org/display/HDF5/H5P_SET_ALIGNMENT):

> Generally the default values will result in the best performance for single-process access to the file.  For MPI IO and other parallel systems, choose an alignment which is a multiple of the disk block size.

Is there a way to explain the difference between your experience and this statement from HDF5?

I have no direct experience with this.  I am just trying to consolidate existing documentation, and make this functionality as clear as possible for netCDF users.  Perhaps the way I rephrased it was not right.

If the issue is complicated and there are doubts about performance behavior, I have no problem completely omitting the performance notes from this function doc.  They could be restored later after the behaviors are better understood.  Your advice?

**Comment by hmaarrfk** [comment]
> Is there a way to explain the difference between your experience and this statement from HDF5?

I have noticed:
1. for non-aligned writes, DIRECT performed terribly. I think I probably had some cache settings set poorly. So badly in fact that I've resorted to opening the file twice. Once for the small bits, the other for the large chunks.
2. I think HDF5 was written at a time where spinning disk drives were the norm. Today, modern SSDs are effectively full processors of their own with their own RAM. The limitations are simply different. The truth is, that up until this year, I wasn't able to measure any difference on a single SSD. 1GB/s was faster than any drive I could write to in a sustained way.

There are also softer advantages to the direct driver.

First, it bypasses the disk cache. For example, if you open and analyze in a serial fashion 100, 1GB files, with the standard SEC2 driver, the Operating system will in effect destroy all its cache by trying to "cache" the 100 files and likely evict more important files. In my experiments, the direct driver avoided the operating system cache alltogether.

This API (and the other I have requested) would allow me to more fully explore the performance axis.

TLDR, I think the old wisdom is centered around HDD. Today, HDDs got faster, and are no-longer the primary drive. (Fast!) SSDs are a paradigm shift that still hasn't made its way to documentation.

**Comment by Dave-Allured** [comment]
> 1. for non-aligned writes, DIRECT performed terribly ...

Okay.  Are you saying that you made a reasonable comparison between **aligned** vs. **non-aligned** writes with **single process** and the DIRECT driver, and **aligned** performed better?

If so, this contradicts the performance statement in the HDF5 doc.  That is not unreasonable.  Performance tuning in general is complicated and a moving target, as you said.

**Comment by hmaarrfk** [comment]
I guess eventually I was comparing:
1. Small bits of unaligned data (my metadata) ~ 200 MB -- Much faster using the default driver (SEC2)
2. Large chunks of aligned data (the main images I was storing). ~40 GB -- Much faster using the Direct Driver

**Comment by Dave-Allured** [comment]
Okay.  Can you easily make a side by side time measurement of large chunks, 5 to 40 Gb with Direct Driver, **single process**, one run with nc_set_alignment or equivalent, and one without?

My goal for today is only to decide what should, and should not, go into the netCDF function description.

**Comment by hmaarrfk** [comment]
| Data size & alignment (RAM)                                                   | HDF5 No Alignment | HDF5 With Alignment |
|-------------------------------------------------------------------------------|-------------------|---------------------|
| 1x 4GB Unaligned                                                              | 2.19 s            | 2.1 s               |
| 1x 4GB Aligned                                                                | 1.87 s            | 1.89 s              |
| 32x 128 MB Unaligned                                                          | 1.67 s            | 1.67 s              |
| 32x 128 MB Aligned                                                            | 1.46 s            | 1.47 s              |
| 32x 128 MB Aligned + DIRECT (https://github.com/Unidata/netcdf-c/issues/2236) | 5.06 s            | 1.01 s              |

Each test is run 4 times taking the average

I then repeated the benchmark running:
```
sudo su -c "echo 1 > /proc/sys/vm/drop_caches"
sudo swapoff -a
sudo swapon -a
```
between each test

| Data size & alignment (RAM)                                                   | HDF5 No Alignment | HDF5 With Alignment |
|-------------------------------------------------------------------------------|-------------------|---------------------|
| 1x 4GB Unaligned                                                              | 2.21 s            | 2.09 s              |
| 1x 4GB Aligned                                                                | 1.92 s            | 1.87 s              |
| 32x 128 MB Unaligned                                                          | 2.17 s            | 2.15 s              |
| 32x 128 MB Aligned                                                            | 1.87 s            | 1.88 s              |
| 32x 128 MB Not Aligned + [DIRECT](https://github.com/Unidata/netcdf-c/issues/2236) | 11.8 s            | 11.8 s              
| 32x 128 MB Aligned + [DIRECT](https://github.com/Unidata/netcdf-c/issues/2236) | 6.48 s            | 1.29 s              |


Benchmarking jupyter notebook
[20220328_hdf5_benchmark.zip](https://github.com/Unidata/netcdf-c/files/8157878/20220328_hdf5_benchmark.zip)


Conclusion: It seems true that there is little benefit for the standard file to enforce alignment with the SEC2 driver.

**Comment by hmaarrfk** [comment]
Hopefully the benchmarks I ran were representative of what you wanted. It is a little hard to run these benchmarks in a repeatable way.

The SSDs get hot, and do their own internal shuffling between different levels of TLC. I've tested this particular SSD up to 2TB of write, then weird stuff happens.

The operating system cache also comes into play with the SEC2 driver (default). Since the SEC2 driver will write to a temporary kernel buffer.

The SSD I got should be fine Sabrent Rocket 4 Plus (4TB) (according to https://www.tomshardware.com/reviews/sabrent-rocket-4-plus-m2-nvme-ssd-review/3) but you know, I could be using it poorly too.

**Comment by hmaarrfk** [comment]
For clarity the intial comparison I was making was between 2.2 s (1.8GB/s) and 1.29 s (3.1GB/s).

**Comment by hmaarrfk** [comment]
An other detail. I found that if i overwrote the same file in my benchmark it would give me strange results.

Therefore, each new write writes to a completely new file. I leave the old files in place.

This is the entry in the fstab
```
/dev/disk/by-uuid/xxxxx /data auto nosuid,nodev,nofail,x-gvfs-show 0 0
```

**Comment by hmaarrfk** [comment]
Output of relevant `mamba list`
```
h5py                      3.6.0           nompi_py39hf407c45_600    ramonaoptics   <---- Includes my patches to expose the direct driver.
hdf5                      1.12.1          nompi_h2750804_103    conda-forge
```
python version 3.9

**Comment by hmaarrfk** [comment]
Really excited that this got through! Thanks you all!
